cmd/traffic/cmd/manager/state/workload_info_watcher.go

```diff
@@ -171,7 +174,60 @@ func rpcWorkloadState(s workload.State) (state rpc.WorkloadInfo_State) {
 	return state
 }
 
-func rpcWorkload(wl k8sapi.Workload, as rpc.WorkloadInfo_AgentState, iClients []*rpc.WorkloadInfo_Intercept) *rpc.WorkloadInfo {
+func extractPortsFromSidecar(_ context.Context, sc *agentconfig.Sidecar) []*rpc.WorkloadPortInfo {
+	if sc == nil {
+		return nil
+	}
+
+	var ports []*rpc.WorkloadPortInfo
+	seenPorts := make(map[types.PortAndProto]bool) // Track seen port+protocol combinations to avoid duplicates
+
+	// Extract ports from all containers in the sidecar
+	for _, container := range sc.Containers {
+		for _, intercept := range container.Intercepts {
+			// Create a unique key using port number and protocol
+			// This allows the same port with different protocols (TCP/UDP)
+			portKey := types.PortAndProto{
+				Port:  intercept.ServicePort,
+				Proto: intercept.Protocol,
+			}
+
+			if !seenPorts[portKey] {
+				seenPorts[portKey] = true
+				protocol := intercept.Protocol.String()
+				ports = append(ports, &rpc.WorkloadPortInfo{
+					Name:     intercept.ServicePortName,
+					Port:     int32(intercept.ServicePort),
+					Protocol: protocol,
+				})
+			}
+		}
+	}
+
+	return ports
+}
+
+func (wf *workloadInfoWatcher) extractPortsForWorkload(ctx context.Context, wl k8sapi.Workload) []*rpc.WorkloadPortInfo {
```

Comment:
This function is works, but it wastes unnecessary resources when a traffic-agent is already present. I suggest adding a check for that. Something like:

```go
  	m := mutator.GetMap(ctx)
  	sc := m.Get(wl.GetName(), wl.GetNamespace())
  	if sc == nil {
```